### PR TITLE
Save functionality and validations implemented

### PR DESCRIPTION
**Pull request by SofiaZarzabal**
- Save functionality implemented
- Empty fields and reservation overlapping validations implemented too

**Folder Structure**
image

**Pixel 3a Demostration**
With a successfully save and a missing field message
PIXEL 3A GIF

**Nexus S Demostration**
With a successfully save and a reservation overlapping message
GIF NEXUS S

**Unit Test**
image

**Unit Test Coverage**
image

### CODE REVIEW

File: app/src/main/java/com/example/parkingsystemkotlin/database/ParkingSpaceReservationDB.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.example.parkingsystemkotlin.entity.Reservation
+
+object ParkingSpaceReservationDB {
+    val hashReservation: MutableMap<Int, MutableList<Reservation>> = HashMap()
+
+    fun getHash(): MutableMap<Int, MutableList<Reservation>> = hashReservation
+
+    fun addReservation(reservation: Reservation) {
+        var reservations: MutableList<Reservation>? = ArrayList()
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystemkotlin/database/ParkingSpaceReservationDB.kt]: pls use the correct colection

File: app/src/main/java/com/example/parkingsystemkotlin/util/ReservationVerify.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        val storedReservationStart = storedReservation.getDateAndTimeStart()
+        val storedReservationEnd = storedReservation.getDateAndTimeEnd()
+
+        if (reservationStart.before(storedReservationStart) && reservationEnd.after(storedReservationStart)) {
+            return true
+        }
+        if (reservationStart.before(storedReservationEnd) && reservationEnd.after(storedReservationEnd)) {
+            return true
+        } else {
+            return reservationStart.after(reservationStart) && reservationEnd.before(storedReservationEnd)
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystemkotlin/util/ReservationVerify.kt]: when condition its better

Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/util/ReservationVerify.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/entity/Reservation.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+class Reservation() {
+    lateinit var dateStart: Calendar
+    lateinit var timeStart: Calendar
+    lateinit var dateEnd: Calendar
+    lateinit var timeEnd: Calendar
+    var parkingSpace: Int = Constants.MINUS_ONE
+    var securityCode: Int = Constants.MINUS_ONE
+
+    fun getReservationVerifyResult(): ReservationVerifiyResult =
+        when {
```
Comment (BQuinteros) [inline app/src/main/java/com/example/parkingsystemkotlin/entity/Reservation.kt:19]: Love it!

Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/entity/Reservation.kt:19]: Thanks!

File: app/src/main/java/com/example/parkingsystemkotlin/activity/ParkingSpaceReservationActivity.kt [hunk 4]
```diff
@@ -38,6 +38,7 @@ class ParkingSpaceReservationActivity : AppCompatActivity(), DatePickerDialog.On
     private fun setListeners() {
         binding.buttonParkingSpaceReservationPickerBegin.setOnClickListener { presenter.onButtonParkingSpaceReservationPickerPressed(this) }
         binding.buttonParkingSpaceReservationPickerEnd.setOnClickListener { presenter.onButtonParkingSpaceReservationPickerPressed(this) }
+        binding.buttonParkingSpaceReservationSave.setOnClickListener { presenter.onButtonParkingSpaceReservationSavePressed() }
```
Comment (FranciscoGaraventa) [inline app/src/main/java/com/example/parkingsystemkotlin/activity/ParkingSpaceReservationActivity.kt]: In this case you could use with(binding){ ... }

File: app/src/main/java/com/example/parkingsystemkotlin/util/ReservationVerify.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import com.example.parkingsystemkotlin.entity.Reservation
+
+class ReservationVerify(var database: ParkingSpaceReservationDB) {
+
+    fun canBeReserved(reservation: Reservation): Boolean {
+        val parkingSpace = reservation.parkingSpace
+        if (database.hashReservation.containsKey(parkingSpace)) {
+            val reservations: MutableList<Reservation>? = database.hashReservation[parkingSpace]
+            reservations?.let {
+                for (r: Reservation in reservations) {
```
Comment (FranciscoGaraventa) [inline app/src/main/java/com/example/parkingsystemkotlin/util/ReservationVerify.kt]: Instead to do this, you can use reservations.forEach { ... }

File: app/src/main/java/com/example/parkingsystemkotlin/database/ParkingSpaceReservationDB.kt (hunk 1 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/database/ParkingSpaceReservationDB.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/util/ReservationVerify.kt (hunk 5 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/util/ReservationVerify.kt]: Done!

File: app/src/main/java/com/example/parkingsystemkotlin/database/ParkingSpaceReservationDB.kt (hunk 1 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/database/ParkingSpaceReservationDB.kt:11]: Done!

**Review by SartiMau — COMMENTED** [review]
Approve after comments solved

File: app/src/main/java/com/example/parkingsystemkotlin/activity/ParkingSpaceReservationActivity.kt (hunk 4 above)
Comment (SofiaZarzabal) [reply app/src/main/java/com/example/parkingsystemkotlin/activity/ParkingSpaceReservationActivity.kt]: Done!

**Review by SartiMau — APPROVED** [review]

**Review by mauroLonghi — APPROVED** [review]

**Review by FranciscoGaraventa — APPROVED** [review]
LG

**Review by BQuinteros — APPROVED** [review]

**Pull request merged**